naga/src/front/wgsl/lower/mod.rs

```diff
@@ -1574,21 +1574,25 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
                         c.ty.map(|ast| self.resolve_ast_type(ast, &mut ectx.as_const()))
                             .transpose()?;
 
-                    let (_ty, init) = self.type_and_init(
+                    let (ty, init) = self.type_and_init(
                         c.name,
                         Some(c.init),
                         explicit_ty,
-                        AbstractRule::Concretize,
+                        AbstractRule::Allow,
                         &mut ectx.as_const(),
                     )?;
                     let init = init.expect("Local const must have init");
 
                     block.extend(emitter.finish(&ctx.function.expressions));
                     ctx.local_table
                         .insert(c.handle, Declared::Const(Typed::Plain(init)));
-                    ctx.named_expressions
-                        .insert(init, (c.name.name.to_string(), c.name.span));
-
+                    // Only add constants of non-abstract types to the named expressions
+                    // to prevent abstract types ending up in the IR.
+                    let is_abstract = ctx.module.types[ty].inner.is_abstract(&ctx.module.types);
+                    if !is_abstract {
+                        ctx.named_expressions
+                            .insert(init, (c.name.name.to_string(), c.name.span));
+                    }
```

Comment:
I think this is another reason why we might not want const expr in named expr at all. https://github.com/gfx-rs/wgpu/issues/6207#issuecomment-2685875562

Reply:
That doesn't seem unreasonable to me.